### ISSUE TITLE
Set initdata in PeerPod 

### ISSUE THREAD

**Issue by huoqifeng** [confidential-containers/cloud-api-adaptor#1895]
This story follows up https://github.com/kata-containers/kata-containers/issues/9468  in cloud-api-adaptor for PeerPod

The initdata schema was discussed, agreed and documented in https://github.com/confidential-containers/trustee/blob/main/kbs/docs/initdata.md, a possible approach is documented in https://github.com/confidential-containers/confidential-containers/issues/171#issuecomment-1922821565

I drew a new diagram to explain more details for https://github.com/confidential-containers/confidential-containers/issues/171#issuecomment-1922821565 In PeerPod, as below:
<img width="840" alt="image" src="https://github.com/confidential-containers/cloud-api-adaptor/assets/15607193/55efd70a-1bc9-4e51-9c8b-f7cc07f38383">

The flow:
1. initdata added in pod descriptor as annotation, example:
```
{
  "algorithm": "sha384",
  "version": "0.1.0",
  "data": {
        "aa.toml": "xxx",
        "cdh.toml": "xxx",
        "certificate.crt": "xxxx",
        "policy.rego": "xxxx",
  }
}
```
2. The annotation passed from Containerd/CRI-O to kata-runtime
3. Generate the initdata files for `kata-cc` and `kata-remote-cc` separately
    - 3.1 for `kata-cc`, the initdata is wrote in pod's specific folder on host
    - 3.2 for `kata-remote-cc`, the initdata passed from kata-runtime to cloud-api-adaptor, cloud-api-adaptor write it to iso or meta-data service. 
4.  Mount the initdata files for `kata-cc` and `kata-remote-cc` separately
    - 4.1 for `kata-cc`, the initdata file will be mounted like via 9p.
    - 4.2 for `kata-remote-cc`, the initdata can be mounted as iso device like libvirt provider
    - 4.3 for `kata-remote-cc`, the initdata can also be retrieved from meta-data service like azure provider
5. initdata format after mount
    Initdata have formats similar as cloud-init yaml but with `"algorithm": "sha384"` and `"version": "0.1.0"` added, so that we can calculate hash based on specific algorithm
6. Provision initdata
    - for `kata-remote-cc`, use `process-user-data` to provision the configurations, like aa.toml, cdh.toml
    - for `kata-cc`, a similar process can be launched by init to do similar but it's not necessary to be a systemd service, it could be a program/scripts in initrd and launched by init 
7. Launch aa, cdh and asr etc
    - Initdata's hash will be added into the evidence
    - Do remote attestation and check the initdata hash on verifier.

I'm expecting only cloud-api-adaptor change for `kata-remote-cc` for this approach. 
Might need some more change in `kata-runtime` for `kata-cc` for this approach. 
cc @Xynnn007 @mkulke @stevenhorsman @fitzthum

**Comment by Xynnn007** [comment]
Wow. Cool diagram, and it is nearly the final shape in my mind.

One thing not clear to me is who processes the initdata toml inside guest?

Another thing to note that, for kata-cc now we are trying to use RPC to inject the initdata into kata-agent directly (under testing) to firstly make it work-able. And the next step would be to think about whether to directly mount the initdata file into guest to a specific path.

If peer-pod is trying to directly put the initdata file into guest, I wonder on kata-cc side, we could directly start discussing and evaluating the way of share the file into guest rather than set with rpc. As the two would meet nearly the same threat model and attack vector.

I'd like to hear more ideas.
cc @fidencio

**Comment by mkulke** [comment]
For context, since `cloud-init` is (too) pervasive and cannot work on a read-only rootfs, @bpradipt wrote `process-user-data` for peerpods, which will just process select `write_files` entries from `cloud-config` yamls.

**Comment by Xynnn007** [comment]
> I assume this has been considered and the RPC approach was preferred, but I don't know really. I can think of one upside when using RPC: dynamic configuration. since you have to pass through the agent to provision files, you can enforce a runtime measurement. Not sure if this is a desirable scenario, but a shared-by-host disk could be read only once.

Yes.I once discussed with @fitzthum and thought the first goal would be "to make it work without changing too many things". But seeing that peerpod is now running quickly to embrace read-only files directly, I begin to wonder whether it is necessary to use RPC on kata first, and then consider using files. Instead, use the file directly (this is cleaner and avoids AA having a strange UpdateConfiguration interface).

Initdata also supports passing policy, so I would also like to hear @danmihai1 's idea.

**Comment by huoqifeng** [comment]
> Some TEEs like SNP (w/o vTPM) cannot add something (extend) the evidence. what they do instead: hash the init-data and add it to guest launch as `HOSTDATA`, which will be part of the signed evidence. the agent then has to assert that `hash(init-data) == HOSTDATA` to bind the init-data to the TEE.

@mkulke could you help please explain more about this? why RPC can resolve the problem but hostpath mount could not?

**Comment by mkulke** [comment]
I am PoC'ing the RPC approach (using only SetPolicy) at the moment, to understand structural implications better.

doing that it feels like there's a lot of ceremony and boilerplate required to add an AA client to the agent... that is in comparison to some host-injected read-only datablob that can be measured into the evidence while kata-agent can be left untouched.

**Comment by mkulke** [comment]
> > Some TEEs like SNP (w/o vTPM) cannot add something (extend) the evidence. what they do instead: hash the init-data and add it to guest launch as `HOSTDATA`, which will be part of the signed evidence. the agent then has to assert that `hash(init-data) == HOSTDATA` to bind the init-data to the TEE.
> 
> @mkulke could you help please explain more about this? why RPC can resolve the problem but hostpath mount could not?

to perform this assertion, you need to fetch an attestation-report. this is done on the guest. The report cannot be injected as part of untrusted host-injected data.

This doesn't really have something to do with RPC, but it happens on the guest. in theory any guest-component (something like `process-user-data`) could do that (hash files; fetch report; perform HOSTDATA check).

**Comment by Xynnn007** [comment]
> I am PoC'ing the RPC approach (using only SetPolicy) at the moment, to understand structural implications better.
> 
> doing that it feels like there's a lot of ceremony and boilerplate required to add an AA client to the agent... that is in comparison to some host-injected read-only datablob that can be measured into the evidence while kata-agent can be left untouched.

Yes. it also makes trouble for launching CDH process, because only after SetInitData is received by kata-agent, CDH can get its configuration and then be launched. This is what Chengyu and I are facing when debugging. If we have a systemd service like `process-user-data` that parses read-only file and place configurations & policies, then AA, CDH, Kata-agent are launched, things would get much simpler.

**Comment by huoqifeng** [comment]
> Yes. it also makes trouble for launching CDH process, because only after SetInitData is received by kata-agent, CDH can get its configuration and then be launched. This is what Chengyu and I are facing when debugging. If we have a systemd service like `process-user-data` that parses read-only file and place configurations & policies, then AA, CDH, Kata-agent are launched, things would get much simpler.

PeerPod is using systemd service, I think it's not necessary to be a systemd service for `kata-cc` if systemd is not enabled. It could just be a program/script in initrd that can be launched by init.

**Comment by mkulke** [comment]
> Yes. it also makes trouble for launching CDH process, because only after SetInitData is received by kata-agent, CDH can get its configuration and then be launched. This is what Chengyu and I are facing when debugging. If we have a systemd service like `process-user-data` that parses read-only file and place configurations & policies, then AA, CDH, Kata-agent are launched, things would get much simpler.

right, sometimes problems in a scheme only become apparent when prototyping an implementation. RPC makes things a bit more complex than I initially assumed. there is a bootstrapping problem also with AA:

We want to receive an AA config via `Agent::SetInitData()`
To bind the InitData to the TEE Evidence we need to either call `AA::ExtendRuntimeMeasurement()` or `AA::GetEvidence()`
We need to have an AA running to call those endpoints

Options:
- launch a half-initialized AA, without the token config, restart AA/reread conf after `SetInitData`. 😐 
- include the `AA::attester` crate as a lib into the agent to call `extend_runtime_measurement()` or `get_evidence()` directly. 🤔 
- Inject a static init-data data blob into the guest. have a one-off guest-component processing this at launch, using the `AA:attester` crate to call `extend_runtime_measurement() or `get_evidence()` to bind it to the TEE evidence. 🤔 🤔

**Comment by huoqifeng** [comment]
> right, sometimes problems in a scheme only become apparent when prototyping an implementation. RPC makes things a bit more complex than I initially assumed. there is a bootstrapping problem also with AA:

So, use `process-user-data` to set initdata sounds like a more straightforward approach and make the API/Implementation more easier. Shall we adopt this approach if it can also bind the HOSTDATA for SNP (w/o vTPM)?

**Comment by fitzthum** [comment]
A few things here.

First, one reason we talked about sending the config via the Kata Agent API is to minimize TCB. There has been a lot of talk about disabling fs passthrough. It's not clear exactly where that will go, but it's something to think about. The RPC option is also more similar to what Dan initially implemented.

I don't really have a preference here but we should think about it carefully. One interesting advantage of decoupling this from the Kata Agent is that it could allow us to use init-data to provision configs for the kata agent. It would probably have implications on how we startup the components in the guest, though.

We have already introduced a `check_init_data` endpoint to the attestation agent. This should be called by whoever processes the init-data. Note that the kata agent is going to be calling the CDH for things anyway. `check_init_data` is designed to be run with the AA in a partially configured state (i.e. the KBS URI has not been provided yet but isn't needed). This means that we also have a somewhat weird flow to update the AA config from the init-data.

The attestation agent also has an `extend_runtime_measurement` endpoint that should be usable in the partially configured state. I guess the caller will have to decide between this and `check_init_data` depending on what platform they are on, which isn't totally ideal.

**Comment by Xynnn007** [comment]
Nice point. One way from my side is to move `check_init_data` API from RPC to launch parameter `--initdata` of AA. The logic inside AA would determine whether to check initdata field, or `extend_runtime_measurement` due to platform. This hints that a common `process-user-data` would pre-process initdata toml into different config/policy files from iso device/cloud metadata service/sharedfs or something.

The key point is the danger of fs passthrough. Are there any details about the talk? A naive idea is that if the dir is read-only, also the guest would never write anything to the shared path, what security flaw else would come?

**Comment by mkulke** [comment]
> The key point is the danger of fs passthrough. Are there any details about the talk? A naive idea is that if the dir is read-only, also the guest would never write anything to the shared path, what security flaw else would come?

I wonder whether consuming init-data from metadata svc, cloud-config iso, agent::SetInitData endpoint are roughly equivalent: in all scenarios we are processing untrusted data from the host in a pre-attested guest environment.

**Comment by huoqifeng** [comment]
I wrote a rough commit to illustrate the approach in PeerPod https://github.com/huoqifeng/cloud-api-adaptor/commit/8fa416d78a7244436c13ce6436eba0e11bebbcd4. It's not done yet but I think we can see the complexity from the commit.

**Comment by fitzthum** [comment]
> The key point is the danger of fs passthrough. Are there any details about the talk? A naive idea is that if the dir is read-only, also the guest would never write anything to the shared path, what security flaw else would come?

I have done some experiments here. Currently it's pretty easy to construct an attack where encrypted container images are leaked through the shared fs. The agent doesn't validate where shared directories are mounted and there are a few ways to create bind mounts inside the guest that can map arbitrary locations (such as the directory where the images are unpacked) onto shared directories. We have done some work to set `fs_share=none`, but currently the host can just re-enable the option without changing the measurement. 

Policy restrictions could be used here, but ideally we would be secure by default without requiring users to have any particular policy. There are also some potential bootstrapping issues if we use the policy to restrict mounts but we need a mount to share the policy. I think the ideal situation would be that the guest simply wouldn't be able to mount any shared directories. For instance, we might be able to disable this at the kernel level. Of course, this would probably break some other things.

One compromise might be to prevent the kata agent from mounting any shared directories but to have one readonly mount specified in fstab. Read-only directories are not guaranteed to be safe. For instance, the kata agent might be tricked into bind mounting `/etc` onto the shared directory. Then the host could overwrite the configuration of the guest. Nonetheless this might be a reasonable compromise.

**Comment by fitzthum** [comment]
> If we would use the aa:attester crate in agent, we'd still have the hw encapsulation (task 1) without the RPC. Is this an option?

This is a really thoughtful suggestion. I think it's probably a good idea. One thing to think about is that we are currently missing a little bit of plumbing to send the init-data to Trustee. When the AA checks the init-data, it should keep track of it so that it can be sent with alongside the evidence. If we are using the AA both as a crate and later as a binary, we might need to store this on the fs somewhere.

Also, would this also work for peer pods or just baremetal?

**Comment by mkulke** [comment]
> When the AA checks the init-data, it should keep track of it so that it can be sent with alongside the evidence.

oh, do we need to attach the full init-data body to the evidence? I assumed it was only the hash which should be HOSTDATA? (or measured inti a PCR, RTMR, ...)

**Comment by fitzthum** [comment]
> oh, do we need to attach the full init-data body to the evidence? I assumed it was only the hash which should be HOSTDATA? (or measured inti a PCR, RTMR, ...)

Trustee is designed to use either one, but one of the ideas behind init-data is that the KBS policy will be able to read the init-data plaintext and make some choices based on it. Ofc if the init-data gets really big (for instance if there is a really big policy file) then this could be a bit clunky. So we're not totally set on this, but we should at least be able to send the plaintext over.

**Comment by arronwy** [comment]
> > The key point is the danger of fs passthrough. Are there any details about the talk? A naive idea is that if the dir is read-only, also the guest would never write anything to the shared path, what security flaw else would come?
> 
> I have done some experiments here. Currently it's pretty easy to construct an attack where encrypted container images are leaked through the shared fs. The agent doesn't validate where shared directories are mounted and there are a few ways to create bind mounts inside the guest that can map arbitrary locations (such as the directory where the images are unpacked) onto shared directories. We have done some work to set `fs_share=none`, but currently the host can just re-enable the option without changing the measurement.
> 

We may need disable virtio-fs/9p in guest kernel which will impact the measurement.

**Comment by arronwy** [comment]
After sync with @Xynnn007 , we found since the integrity of initdata passed from host side is ensured by TEE HW register, we can send initdata to guest with below four options: shared fs/shared volume/RPC/network. 

But from security point of view as @fitzthum mention, shared fs (virtio-9p/fs) is disabled for enhanced security, 

RPC way may need a new service daemon or reuse `kata-agent` which block the init config for `kata-agent` it self. 

Network way may need an dedicated `metadata-service` as current peer pod use.

The shared volume way maybe a ideal option, it is easy to use, `containerd-shim-kata-v2` will create a initdata block device and share it to guest through `virtio-blk`, inside guest, we can use systemd service mount this device to `/mn/init_data` directory before kata-agent is launched and for other service daemon to consume, we can write a separate program to parse the init data or depends on kata-agent to do the parse job.

**Comment by mkulke** [comment]
One thing to consider, which we briefly discussed on the peerpod community call would be size-limits of user-data bodies: AWS, Azure, GCP all have various limits here, varying between 16kb - 256kb. I'm not sure whether that's prohibitive. a cert is alone is >1kb, a policy body could maybe be too large (when autogenerated from a complex k8s manifest).

**Comment by huoqifeng** [comment]
<img width="724" alt="image" src="https://github.com/confidential-containers/cloud-api-adaptor/assets/15607193/179a973c-96b2-4202-8107-75bbb4fe4470">

To resolve the issue that policy size might large than the user-data size limitation, this might be a solution after discussion with @stevenhorsman

We set initdata and pass remote attestation as illustrated in the diagram above (Phase 1), and we have a SetPolicy API (it exists today and will continue exist after implemented initdata API). In the SetPolicy API within guest, we retrieve the policy's hash value from KBS and make it happen only if the hash matches (Phase 2). In this approach we can also support SetPolicy multiple times safely by change it's hash in KBS.  

If policy file is not too large, it can be set together with initdata also.
@mkulke @stevenhorsman @Xynnn007 @arronwy @fitzthum

**Comment by mkulke** [comment]
to get some numbers I checked the output from kata's genpolicy tool for a simple 1-container nginx deployment. so yes, in practice those would get pretty large.

```
wc -c <(base64 -w0 ../my-policy.rego)
70160 /dev/fd/63
```

**Comment by mkulke** [comment]
> We set initdata and pass remote attestation as illustrated in the diagram above (Phase 1), and we have a SetPolicy API (it exists today and will continue exist after implemented initdata API). In the SetPolicy API within guest, we retrieve the policy's hash value from KBS and make it happen only if the hash matches (Phase 2). In this approach we can also support SetPolicy multiple times safely by change it's hash in KBS.

hmm to me it feels odd to use KBS as a digest store for integrity verifications (we already kind of have this for signatures, so it's not unthinkable I guess).

It would also impose a burden to the user to provision the digests, but only if the policy body is too large, so there's are 2 options to choose from.

I would lean towards the least complex option from the architecture and user experience. At the moment I feel a SetInitData RPC with a flat body to provision all files would probably be a simpler solution, even if it requires more plumbing internally (re-configuring AA in-flight)

**Comment by huoqifeng** [comment]
I just knew the policy file is more than `1700` lines if it's generated from genpolicy tool, like https://github.com/kata-containers/kata-containers/blob/main/src/tools/genpolicy/samples/pod-one-container.yaml#L9, I doubt whether it's reasonable for a user to understand it before use it. This might be another topic but I think it's very important to simplify the policy file itself if we want user use it in real case.

**Comment by fitzthum** [comment]
If the policy or other init-data is too big to send to the KBS, we don't need to get the hash from the KBS. We can just use the existing design, but not send over the init-data. Remember that the policy is measured into the host data aka hw init-data (on baremetal) or measured by the TPM (peer pods). So the hash of the policy is going to be in the evidence either way.

If the init-data is too large to send, the KBS can be provisioned with expected hashes and can use those to validate the evidence. Of course this doesn't allow the KBS to examine the contents of the init-data, which is a shame, but it is a workable backup option.

In previous discussion of init-data we have agree that there is no need to set the init-data more than once. Let's not worry about that. It makes things more complicated.

I do think it's a problem that the policy is so large and complex, but remember that the genpolicy tool is just one way to generate a policy.

**Comment by arronwy** [comment]
Kubernetes annotation size can up to 256 KB: https://github.com/kubernetes/kubernetes/pull/16068/files
Kubernetes config map can up to 1MB: https://kubernetes.io/docs/concepts/configuration/configmap/#motivation

If policy file is too big to set through K8S, it can be defined as a CDH resource and get from remote KBS during CDH initialization which current CDH already support this feature, the only issue is we may need change the boot sequence from `kata-agent->aa->CDH` to `aa->CDH->kata-agent` @Xynnn007 @fitzthum @stevenhorsman @huoqifeng

**Comment by huoqifeng** [comment]
I think `aa->CDH->kata-agent` is reasonable as it reflects the functions dependency.

**Comment by Xynnn007** [comment]
This hints that we need systemd, even for initrd

**Comment by mkulke** [comment]
we might be following different threads here...

> I'm not sure that the KBS protocol actually has any size limits. It seems like the problem you're running into is that the method of getting the init-data plaintext into a cloud guest usually has limits, but I think that's mostly unrelated to what you do with the init-data once it's inside the guest.

indeed. AFAIU the discussion is about whether we can **replace an SetInitData RPC** call with a simpler way of provisioning config policy, e.g. via **read-only config disks** or a call to a cloud's **metadata endpoint (IMDS)**. The latter has size limits that are prohibitive if we want to provision policies (the limit for AWS is 16kb).

So, I guess we can conclude, IMDS should not be the default mechanism to provision a policy.

The follow up questions:
- Should we still attempt to provision init-data via config disks or IMDS. w/o a policy body?
- Can we still use `SetPolicy()`? (extend it with `verify_policy_hash/extend_runtime_measurement`)
- Should the policy be retrieved from somewhere else?
- Does it make sense for KBS to manage agent policies? it feels like we're diluting the notion of a confidential resource

**Comment by huoqifeng** [comment]
I made a PoC for this approach in PeerPod for libvirt provider with initdata which includes aa.toml, cdh.toml and a small policy file in PR (branch `4.2` in the diagram):
https://github.com/confidential-containers/cloud-api-adaptor/pull/1912
https://github.com/kata-containers/kata-containers/pull/9988

**Comment by fitzthum** [comment]
Pushing the init-data from the KBS->guest does have the benefit of supporting platforms that don't have hw init data fields or TPM support. On the other hand it has some annoying bootstrapping issues given that the AA needs to be fully configured to be able to get the resource from the KBS. I'm not really a fan of this approach.

It seems like we might be headed towards using block devices rather than RPC to provision the init-data on baremetal. Maybe that is a decent parallel to so-called config disks. This might imply creating another component and requiring systemd, but could we just handle things in the Kata Agent instead?

**Comment by mkulke** [comment]
> This might imply creating another component and requiring systemd, but could we just handle things in the Kata Agent instead?

I don't think it requires systemd. that would help with the orchestration of the individual guest components, but I don't think we would necessarily need an additional component. instead of being passively poked on a SetInitData(), the agent would actively call GetInitData() and perform the same things. But yeah, if we need to provision a larger blob of data from the runtime to the guest RPC might be just the right facility.

**Comment by huoqifeng** [comment]
The main function has been committed in PR https://github.com/confidential-containers/cloud-api-adaptor/pull/1912, I'm reopening this story to track remain issues:
- [ ] Replace AA_KBC_PARAMS after enable initdata https://github.com/confidential-containers/cloud-api-adaptor/issues/1985
- [ ] A potential malicious use case that a container might escape from rust-jail, flow:
- A malicious initdata is generated with a danger policy in it
- The malicious initdata is used by a TEE guest and a malicious container tampered the digest after escaped rust-jail (in case it know a good policy and know a good initdata's digest)
- The malicious container leverage the tampered digest and passed the attestation
- The malicious container got a secret from KBS
I think a solution might be as described in `step 5, 6, 7` in https://github.com/confidential-containers/confidential-containers/issues/171#issuecomment-1922821565. Maybe, we can add this logic either in process-user-data or in kata-agent to do attestation before kata-agent receive any requests.
CC @mkulke @Xynnn007 @fitzthum

**Comment by mkulke** [comment]
> a malicious container tampered the digest after escaped rust-jail (in case it know a good policy and know a good initdata's digest)

It might be worth pointing out that this threat vector only exists if we rely on software measurements in userland. HOSTDATA, RTMRs, PCRs cannot be arbitrarily edited from the guest.

> (in case it know a good policy and know a good initdata's digest)

I think we can assume this is the case. initdata (and hence it's digest) is not a secret.

> I think a solution might be as described

I assume the suggestion is to perform an initial remote-attestation (e.g. get kbs token w/o any secret release?) before a container would be launched. is this correct?

**Comment by huoqifeng** [comment]
> I assume the suggestion is to perform an initial remote-attestation (e.g. get kbs token w/o any secret release?) before a container would be launched. is this correct?

Yes, initial remote-attestation does not try to retrieve any secrets but just make sure the initdata's integrity. Guest should exit/stop if initial remote attestation fails.

**Comment by fitzthum** [comment]
@huoqifeng so SE has no equivalent to the hostdata or PCRs? Without a field that is isolated from the guest, I think you're going to have some inevitable risks. Doing an initial attestation might make it a little bit harder for an attacker, but I don't think it does much. 

A malicious host could simply reroute the initial attestation to their own KBS. Once a container is running, if they break out of it into the guest, they can do additional attestations with whatever report data they want. It doesn't seem like there is anything that links the first attestation to subsequent ones.

**Comment by huoqifeng** [comment]
@fitzthum malicious KBS is mentioned in `step 8` in https://github.com/confidential-containers/confidential-containers/issues/171#issuecomment-1922821565, we do need additional settings like network/firewall rules to prevent malicious KBS.
I/m not aware of SE has equivalent to the hostdata or PCRs, but will confirm with hardware/firmware friends. We may also have some other options to prevent this attach, like:
- Use encrypted OCI image
- Use signed OCI image
So that container can not be launched at all if attestation failed.

**Comment by fitzthum** [comment]
@huoqifeng Let's say that you have a workload that uses only encrypted container images. Let's say I am a malicious host. Before running your workload. I can start my own workload. Since this is my workload I can point it to my KBS and run whatever containers I want. So I pick a container image that allows me to connect to it and I figure out how to break out of that container image into the guest.

Now I simply take the initdata that your workload would have started with and I put it into a report generated in my guest. Now I have evidence that I can send to your KBS that will look the same as if it came from your guest. With this evidence, your KBS will send me the keys to decrypt your image.

**Comment by huoqifeng** [comment]
@fitzthum , malicious KBS is the key for this threat, I'm wondering how digest based on hostdata or PCRs prevent this attack for this case?

**Comment by mkulke** [comment]
using a malicious kbs would be a way to work around a protection scheme based on initial remote-attestation/encrypted or signed images, since we could pivot and request secrets from a real kbs with forged evidence later, after such a step.

the difference to a pure software measurement log would be: for HOSTDATA or RTMRs/PCRs evidence cannot be forged, it's either simply a static read-only part of a signed hw-report or it's an extend-only interface for the latter two. when you have an initial measurement M, and want to turn it into measurement N, you'd need to find X in N = hash(M || hash(X)), which should not be possible due to hash being a one-directional function.

**Comment by fitzthum** [comment]
> malicious KBS is the key for this threat

In my example there is really no malicious KBS. The attacker starts a workload pointing to their own KBS, which is completely valid.

**Comment by huoqifeng** [comment]
I'm still confused here, if I am a malicious host and I have a container that can escape from rust_jail, why not use your initdata directly to run my container to get the secret? I guess as a host I can run it in the real TEE with my malicious container.

**Comment by fitzthum** [comment]
> I'm still confused here, if I am a malicious host and I have a container that can escape from rust_jail, why not use your initdata directly to run my container to get the secret? I guess as a host I can run it in the real TEE with my malicious container.

I'm assuming the target workload is configured with initdata that only allows certain containers to be run (with signatures or policy).  So if the host provides this valid init data at boot, then those containers must be the ones that are started.

But if the host launches a completely different workload, they can specify init data that allows their own container to run. Then they connect to the container, break into the host, and request another hw report that has your valid init data in it rather than their own. This is the evidence that can be used to get your secrets.

**Comment by huoqifeng** [comment]
> > I'm still confused here, if I am a malicious host and I have a container that can escape from rust_jail, why not use your initdata directly to run my container to get the secret? I guess as a host I can run it in the real TEE with my malicious container.
> 
> I'm assuming the target workload is configured with initdata that only allows certain containers to be run (with signatures or policy). So if the host provides this valid init data at boot, then those containers must be the ones that are started.
> 
> But if the host launches a completely different workload, they can specify init data that allows their own container to run. Then they connect to the container, break into the host, and request another hw report that has your valid init data in it rather than their own. This report will say that the guest was launched with your init data but it was not and the host can use this evidence to get your secrets.

Makes sense, sounds like this is something role based secrets release KBS should support. I guess it's not supported yet, is there a thread talking about this?

**Comment by huoqifeng** [comment]
I'll open a new story to track potential vulnerability issue described in the thread and close this story.
